src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentitySource.cs

```diff
@@ -0,0 +1,90 @@
+    {
+        internal const string AuthenticationResponseInvalidFormatError = "Invalid response, the authentication response was not in the expected format.";
+        internal const string UnexpectedResponse = "Managed Identity response was not in the expected format. See the inner exception for details.";
+        internal RequestContext _requestContext;
+
+        protected ManagedIdentitySource(RequestContext requestContext)
+        {
+            _requestContext = requestContext;
+        }
+
+        protected internal string ClientId { get; }
+
+        public virtual async Task<ManagedIdentityResponse> AuthenticateAsync(AppTokenProviderParameters parameters, CancellationToken cancellationToken)
+        {
+            ManagedIdentityRequest request = CreateRequest(parameters.Scopes.ToArray());
+            
+            var response =
+            HttpMethod.Get.Equals(request.Method) ?
+            await _requestContext.ServiceBundle.HttpManager.SendGetAsync(request.UriBuilder.Uri, request.Headers, _requestContext.Logger, isManagedIdentity: true, cancellationToken: cancellationToken).ConfigureAwait(false) :
+            await _requestContext.ServiceBundle.HttpManager.SendPostAsync(request.UriBuilder.Uri, request.Headers, request.BodyParams, _requestContext.Logger, isManagedIdentity: true, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+            return HandleResponse(parameters, response);
+            
+        }
+
+        protected virtual ManagedIdentityResponse HandleResponse(
+            AppTokenProviderParameters parameters,
+            HttpResponse response)
+        {
+            try
+            {
+                if (response.StatusCode == System.Net.HttpStatusCode.OK)
+                {
+                    return GetSuccessfulResponse(response);
+                }
+
+                throw MsalServiceExceptionFactory.FromManagedIdentityResponse(MsalError.ManagedIdentityFailedResponse, response);
+            }
+            catch (Exception e)
+            {
+                if (e is MsalServiceException)
+                    throw;
+
+                throw new MsalServiceException(MsalError.UnknownManagedIdentityError, UnexpectedResponse, e);
+            }
+        }
+
+        protected abstract ManagedIdentityRequest CreateRequest(string[] scopes);
+
+
+        protected string GetMessageFromResponse(HttpResponse response)
+        {
+            if (response.Body.IsNullOrEmpty())
+            {
+                _requestContext.Logger.Info("The response body is empty.");
+                return null;
+            }
+
+            return JsonHelper.DeserializeFromJson<ManagedIdentityErrorResponse>(response.Body).Message;
+        }
+
+        protected ManagedIdentityResponse GetSuccessfulResponse(HttpResponse response)
+        {
+            ManagedIdentityResponse managedIdentityResponse = JsonHelper.DeserializeFromJson<ManagedIdentityResponse>(response.Body);
```

Comment:
What if `response.Body` is null?

Reply:
When the response body is null it is captured in the if block [below](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/nebharg%2FMsiPrototype/src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentitySource.cs#L75) and throws MsalServiceException. There is a [unit test](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/nebharg%2FMsiPrototype/tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/AppServiceTests.cs#L107) to test this scenario as well.